docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,293 @@
+          apiServerEndpoint: {{ index .MasterIPs 0 }}:443
+      nodeRegistration:
+        taints: []
+        kubeletExtraArgs:
+          node-labels: ""
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+  trustd:
+    token: '{{ .TrustdInfo.Token }}'
+    endpoints: [ {{ .Endpoints }} ]
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+
+## Proposal
+
+As you can see, there are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+
+```
+#!talos
+version: "v0.1.0"
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+    token: abc.1234
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+    token: abc.1234
+machine:
+  nodeType: init, controlplane, worker
+  masters: [x.x.x.x, y.y.y.y, z.z.z.z]
+  kubeletExtraArgs: {}
+cluster:
+  clusterName: {{ .ClusterName }}
+  controlPlane:
+    apiServerExtraArgs: {}
+    controllerManagerExtraArgs: {}
+    schedulerExtraArgs: {}
+    networking:
+      dnsDomain: {{ .ServiceDomain }}
+      podSubnet: {{ index .PodNet 0 }}
+      serviceSubnet: {{ index .ServiceNet 0 }}
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+I'm hoping that most kubernetes-specific knobs that need to be tweaked can be done with extra args. The machine and cluster top-level keys are new and correspond to things that need to be configured per-machine, as well as things that need to be set at the cluster level. The install and networking layouts will likely change as other proposals are implemented in Talos v0.2.0 so those are left unchanged for now.
```

Comment:
Aside from the machine config layout, I think I'd like to see something around versioning and we intend on implementing backwards compatibility, config verification, config translation, how far back we will support a config, etc.

Reply:
pre-v1.0, we'll support n-1 on versions for machine config.